src/webview/java/org/medicmobile/webapp/mobile/EmbeddedBrowserActivity.java

```diff
@@ -300,6 +312,36 @@ public void errorToJsConsole(String message, Object... extras) {
 	}
 
 //> PRIVATE HELPERS
+	private String requestCodeToString(int requestCode) {
+		if (requestCode == ACCESS_FINE_LOCATION_PERMISSION_REQUEST_CODE) {
+			return "ACCESS_FINE_LOCATION_PERMISSION_REQUEST_CODE";
+		}
+		if (requestCode == DISCLOSURE_LOCATION_ACTIVITY_REQUEST_CODE) {
+			return "DISCLOSURE_LOCATION_ACTIVITY_REQUEST_CODE";
+		}
+		return String.valueOf(requestCode);
+	}
+
+	private void processRdToolkitProvisionTestActivity(int requestCode, int resultCode, Intent intentData) {
+		String provisionScript = rdToolkitSupport.process(requestCode, resultCode, intentData);
+		trace(this, "RDToolkitSupport :: Executing JavaScript: %s", provisionScript);
+		evaluateJavascript(provisionScript);
+	}
+
+	private void processRdToolkitCaptureResultActivity(int requestCode, int resultCode, Intent intentData) {
+		int readPermission = ContextCompat.checkSelfPermission(this, READ_EXTERNAL_STORAGE);
+
+		if (readPermission != PERMISSION_GRANTED) {
+			trace(this, "RDToolkitSupport :: Requesting storage permissions");
+			ActivityCompat.requestPermissions(this, PERMISSIONS_STORAGE, ACCESS_STORAGE_PERMISSION_REQUEST_CODE);
+			return;
```

Comment:
So, we return here if permission isn't granted and we request it. Does this mean that the capture script that follows never runs now? Or is this code somehow triggered again? 

If not, could we extract the capture script (the rest of this function) in a separate function and maybe also run it when `ACCESS_STORAGE_PERMISSION_REQUEST_CODE` is received?

Reply:
The idea of user experience is:
- The user accepts the permission
- then click in "capture" button. That's why no action [here](https://github.com/medic/medic-android/pull/175/files/6776128ea7d3a4a59102c0837d9d96a704f851c8#r637007069) 
- medic-android closes 
- RDtoolkit app opens and request the user to capture results

When testing I didn't feel it was a weird user experience and I feel it gives a transition experience when switching apps but I'm the only one using this so far and not the target user so feedback on this very valuable. 

What do you think? Better to trigger the action automatically after the permission is granted? The experience would be: 
- The user accept the permission
- Immediately medic-android closes
- RDToolkit app opens and request the user to capture the results.

Reply:
But the permission isn't requested every time, right? If the permission is already granted, we're already in the 2nd situation.

Reply:
I'm also unsure what triggers the "RDTOOLKIT_CAPTURE_ACTIVITY_REQUEST_CODE". Is it the user clicking on the "capture" button?

Reply:
> But the permission isn't requested every time, right? 

If the app already has the permission granted then it wont ask for it again. The experience is when clicking the capture button then the RDToolkit app opens (no modal asking for permission before opening RDToolkit) 

> I'm also unsure what triggers the "RDTOOLKIT_CAPTURE_ACTIVITY_REQUEST_CODE". Is it the user clicking on the "capture" button in the first place? If there's a permission request, then the user needs to click capture twice?

Yes, if the permission isn't granted, then click capture button, modal appears, accept permission, and click capture button, RDToolkit open. 

We reach this android code when clicking on the button, I can try checking the permission somewhere else (tEmbeddedBrowserActivity > onStart() maybe ?), probably the modal asking for permissions will display when opening the medic-android (I guess ?)  
Maybe this is better experience and will cover any other feature needing the same permission 🤔

Reply:
I think we should be consistent about how the app behaves with or without the permission. It's not intuitive that you need to click the button the 2nd time if you didn't have permission.

Reply:
So we should trigger the "capture" part when the permission is granted, like in my initial comment maybe?

Reply:
Fixed